plugin/federation/federation.go

```diff
@@ -281,67 +293,69 @@ func (f *federation) setEntities(schema *ast.Schema) {
+					resolverName = fmt.Sprintf("findMany%s", resolverFieldsToGo)
+				} else {
+					resolverName = fmt.Sprintf("find%s", resolverFieldsToGo)
 				}
 
-				// Let's process custom entity resolver settings.
-				dir = schemaType.Directives.ForName("entityResolver")
-				if dir != nil {
-					if dirArg := dir.Arguments.ForName("multi"); dirArg != nil {
-						if dirVal, err := dirArg.Value.Value(nil); err == nil {
-							e.Multi = dirVal.(bool)
-						}
-					}
-				}
-
-				// If our schema has a field with a type defined in
-				// another service, then we need to define an "empty
-				// extend" of that type in this service, so this service
-				// knows what the type is like.  But the graphql-server
-				// will never ask us to actually resolve this "empty
-				// extend", so we don't require a resolver function for
-				// it.  (Well, it will never ask in practice; it's
-				// unclear whether the spec guarantees this.  See
-				// https://github.com/apollographql/apollo-server/issues/3852
-				// ).  Example:
-				//    type MyType {
-				//       myvar: TypeDefinedInOtherService
-				//    }
-				//    // Federation needs this type, but
-				//    // it doesn't need a resolver for it!
-				//    extend TypeDefinedInOtherService @key(fields: "id") {
-				//       id: ID @external
-				//    }
-				if !e.allFieldsAreExternal() {
-					resolverFields := []string{}
-					for _, f := range e.KeyFields {
-						resolverFields = append(resolverFields, f.Field.ToGo())
-					}
-
-					resolverFieldsToGo := schemaType.Name + "By" + strings.Join(resolverFields, "And")
-					if e.Multi {
-						resolverFieldsToGo += "s" // Pluralize for better API readability
-						e.ResolverName = fmt.Sprintf("findMany%s", resolverFieldsToGo)
-					} else {
-						e.ResolverName = fmt.Sprintf("find%s", resolverFieldsToGo)
-					}
+				e.Resolvers = append(e.Resolvers, &EntityResolver{
+					ResolverName: resolverName,
+					KeyFields:    keyFields,
+					InputType:    resolverFieldsToGo + "Input",
+				})
+			}
 
-					e.InputType = resolverFieldsToGo + "Input"
+			e.Requires = []*Requires{}
+			for _, f := range schemaType.Fields {
+				dir := f.Directives.ForName("requires")
+				if dir == nil {
+					continue
+				}
+				if len(dir.Arguments) != 1 || dir.Arguments[0].Name != "fields" {
+					panic("Exactly one `fields` argument needed for @requires declaration.")
+				}
+				requiresFieldSet := fieldset.New(dir.Arguments[0].Value.Raw, nil)
+				for _, field := range requiresFieldSet {
+					e.Requires = append(e.Requires, &Requires{
+						Name:  field.ToGoPrivate(),
+						Field: field,
+					})
 				}
-
-				f.Entities = append(f.Entities, e)
 			}
 		}
+		f.Entities = append(f.Entities, e)
 	}
 
 	// make sure order remains stable across multiple builds
 	sort.Slice(f.Entities, func(i, j int) bool {
 		return f.Entities[i].Name < f.Entities[j].Name
 	})
 }
+
+func isFederatedEntity(schemaType *ast.Definition) ([]*ast.Directive, bool) {
```

Comment:
Ah thank you!  Yeah this is is great.  We can tweak the function name as we use it more.